src/storage/AdminProcessor.h

```diff
@@ -41,14 +41,48 @@ class TransLeaderProcessor : public BaseProcessor<cpp2::AdminExecResp> {
         auto host = kvstore::NebulaStore::getRaftAddr(HostAddr(req.get_new_leader().get_ip(),
                                                                req.get_new_leader().get_port()));
         part->asyncTransferLeader(host,
-                                  [this, spaceId, partId] (kvstore::ResultCode code) {
-            auto leaderRet = kvstore_->partLeader(spaceId, partId);
-            CHECK(ok(leaderRet));
+                                  [this, spaceId, partId, part] (kvstore::ResultCode code) {
             if (code == kvstore::ResultCode::ERR_LEADER_CHANGED) {
-                auto leader = value(std::move(leaderRet));
-                this->pushResultCode(to(code), partId, leader);
+                LOG(INFO) << "I am not the leader yet!";
+                handleLeaderChanged(spaceId, partId);
+                onFinished();
+                return;
+            } else if (code == kvstore::ResultCode::SUCCEEDED) {
+                // To avoid dead lock, we use another ioThreadPool to check the leader information.
+                folly::via(folly::getIOExecutor().get(), [this, part, spaceId, partId] {
+                    int retry = FLAGS_waiting_new_leader_retry_times;
+                    while (retry-- > 0) {
+                        auto leaderRet = kvstore_->partLeader(spaceId, partId);
+                        if (!ok(leaderRet)) {
+                            this->pushResultCode(to(error(leaderRet)), partId);
+                            onFinished();
+                            return;
+                        }
+                        auto leader = value(std::move(leaderRet));
+                        auto* store = static_cast<kvstore::NebulaStore*>(kvstore_);
+                        if (leader != HostAddr(0, 0) && leader != store->address()) {
```

Comment:
Why not check leader is same with target leader?

Reply:
Actually, we don't need to do it right now. Because the only thing we care about is the leader is not current host.